### DISCUSSION TITLE
Question for real-time performances (Python vs C++)

### DISCUSSION THREAD

**Discussion by sholybest** [loco-3d/crocoddyl#1241] [Q&A]
Thank you to the development team for providing this wonderful library and to everyone reading this message.
I'm going to use the Box-FDDP solver for real-time quadruped MPC simulation.
My desktop CPU is Intel i5-6600k.

When I run the example code "quadrupedal_walking_ubound.py", solver.solve() takes about 0.2 seconds. Although my hardware is not a good one, this is too slow for real-time implementation. So I would like to make a C++ version of the same problem.

But as I know, crocoddyl is compiled from C++ and the python code calls C++ functions when solving the problem.
I think the only additional overload comes from calling the C++ function. If the function is a heavy one, which does many calcuations, then the overload from calling might not have a big effect..

Is there any dramatic change in performance when the solver is transferred into C++?
And is there any reference materials for converting the code to C++?

Thank you.

**Comment by cmastalli** [comment]
Hi @sholybest,

The C++ implementation would certainly improve a bit the computation time, but not substantially. The solver, dynamics, costs, etc are already written in C++. You are indeed using C++ code for most of the job, except formulating the problem (not resolving it).

You must consider a few aspects to improve computation:
  1. Compile Crocoddyl with multi-threading support (check cmake variable).
  2. Compile Crocoddyl and its dependencies for your native architecture (check how to set the compilation flag on the internet).
  3. Do not solve the entire problem online, as done in MPC approaches (e.g., https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=10101674).
  4. Code-generate the dynamics (this is not supported in Python ATM).

Additionally, we're working on a much more efficient solver. However, you'll need to be patient as this won't be released now.

Best,
Carlos

**Reply by sholybest** [reply]
Thank you very much for such a detailed answer. I'll looking forward the new releases!!
